File: plugins/priorities/tutorial/pev_2_verify.cpp
// @file pev_3_verify.cpp verification with std synchronous marking

// 1) generate models via "pev_cbs_setup.lua"
// 2) run this executable


#include "libfaudes.h"

using namespace faudes;



// main
int main(int argc, char* argv[]) {

    ConsoleOut::G()->Verb(1);
    GeneratorVector gvoi;
    gvoi.Clear();
    EventPriorities prios;
    
    // number of conveyor belts (needs to match cbs_setup.lua)
    const int count = 8;

    // read models
    gvoi.Append(Generator("tmp_pev_cbs_src.gen"));
    for(int i = 1;i<=count;i++){
        gvoi.Append(Generator("tmp_pev_cbs_"+ToStringInteger(i)+"_cl.gen"));
    }
    gvoi.Append(Generator("tmp_pev_cbs_snk.gen"));

    // read priorities
    prios.Read("tmp_pev_cbs_prios.alph");

    // do compositional verification
    auto start = std::clock();
    bool isnc = IsPNonblocking(gvoi,prios);
    std::cout<<"Duration in seconds: "<<ToStringFloat((std::clock()-start)/(double) CLOCKS_PER_SEC)<<std::endl;
    std::cout<<"Is P-Nonconflicting? "<< isnc<<std::endl;

    // do monolytic verification
    start = std::clock();
    Generator gall;
    gall.StateNamesEnabled(false);
    ParallelNB(gvoi,gall);
    ShapePriorities(gall,prios);
    isnc = IsNonblocking(gall);
    std::cout<<"Duration in seconds: "<<ToStringFloat((std::clock()-start)/(double) CLOCKS_PER_SEC)<<std::endl;
    std::cout<<"Is P-Nonconflicting? "<< isnc<<std::endl;

    return 0;


}

File: plugins/priorities/tutorial/pev_2_verify.lua
-- Conveyor Belts Demo for Compositional Verification
--
-- Uses "pev_cbs_setup.lua" to generate the models


-- configure number of conveyor belts via command line, default to 5
N=arg[1]
if N == nil then N=5 end  
print(string.format('pev_2_verify.lua: number of conveyor belts N=%d',N))

-- generate models
dofile("pev_cbs_setup.lua")

-- read models
print('pev_2_verify.lua: reading models')
gvec=faudes.GeneratorVector()
gvec:Append('tmp_pev_cbs_src.gen')
for i=1,N do
  gvec:Append(string.format('tmp_pev_cbs_%d_cl.gen',i))
end
gvec:Append("tmp_pev_cbs_snk.gen")

-- read priorities
prios = faudes.EventPriorities("tmp_pev_cbs_prios.alph");

-- do verify (compositional)
print('pev_2_verify.lua: compositional verification')
isnc = faudes.IsPNonblocking(gvec,prios)
if isnc then
  print('pev_2_verify: nonconflictingness confirmed (PASS)')
else
  print('pev_2_verify: nonconflictingness failed (FAIL)')
end
FAUDES_TEST_DUMP("prio comp ver",isnc)

-- do verify (monolitic, for limited number of conveyor belts)
if tonumber(N) <6 then
  print('pev_2_verify.lua: monolitic verification')
  gall=faudes.Generator()
  faudes.ParallelLive(gvec,gall)
  print('pev_2_verify: overall statecount:',gall:Size())
  faudes.ShapePriorities(gall,prios)
  isnc = faudes.IsNonblocking(gall)
  if isnc then
    print('pev_2_verify: nonconflictingness confirmed (PASS)')
  else
    print('pev_2_verify: nonconflictingness failed (FAIL)')
  end
end

